Validation/HcalRecHits/src/HcalRecHitsValidation.cc

```diff
@@ -387,8 +400,95 @@ void HcalRecHitsValidation::analyze(edm::Event const& ev, edm::EventSetup const&
 
     
   }
-  //  std::cout << "*** 9" << std::endl; 
 
+  //SimHits vs. RecHits
+
+  if(subdet_ > 0 && subdet_ < 6 && imc !=0) {  // not noise 
+
+    edm::Handle<PCaloHitContainer> hcalHits;
+    ev.getByToken(tok_hh_,hcalHits);
+    const PCaloHitContainer * SimHitResult = hcalHits.product () ;
+    
+    double enSimHits    = 0.;
+    double enSimHitsHB  = 0.;
+    double enSimHitsHE  = 0.;
+    double enSimHitsHO  = 0.;
+    double enSimHitsHF  = 0.;
+    double enSimHitsHFL = 0.;
+    double enSimHitsHFS = 0.;
+    // sum of SimHits in the cone 
+    
+    for (std::vector<PCaloHit>::const_iterator SimHits = SimHitResult->begin () ; SimHits != SimHitResult->end(); ++SimHits) {
+
+      int sub, depth;
+      HcalDetId cell; 
+
+      if (testNumber_) {
+        int z, lay, eta, phi;
+        HcalTestNumbering::unpackHcalIndex(SimHits->id(), sub, z, depth, eta, phi, lay);
+        int sign = (z==0) ? (-1):(1);
+        eta     *= sign;
+        //The id used in testnumbering encodes the cell information, we generate a HcalDetId that we can then use to extract eta&phi (not ieta and iphi) information from.
+        // HO is handled differently, since it doesn't have a depth 1 defined.
+        if(sub == static_cast<int>(HcalOuter)){
+          cell = HcalDetId(HcalSubdetector::HcalOuter,eta,phi,4);
+        } else {
+          cell = HcalDetId(static_cast<HcalSubdetector>(sub),eta,phi,1);
+        }
+
+      } else {
+
+        cell         = HcalDetId(SimHits->id());
+        sub          = cell.subdet();
+        depth        = cell.depth();
+      }
+
+      const CaloCellGeometry* cellGeometry =
+	geometry->getSubdetectorGeometry (cell)->getGeometry (cell) ;
+      double etaS = cellGeometry->getPosition().eta () ;
+      double phiS = cellGeometry->getPosition().phi () ;
+      double en   = SimHits->energy();    
+
+      double r  = dR(eta_MC, phi_MC, etaS, phiS);
+       
+      if ( r < partR ){ // just energy in the small cone
+	enSimHits += en;
+	if(sub == static_cast<int>(HcalBarrel)) enSimHitsHB += en; 
```

Comment:
HF has to be handled separately regardless since HF simhit energy is defined in p.e.'s and not GeV.

I think that it still makes sense to compare SimHit to RecHit energy for each subdetector, especially when their are different electronics in HB and HE.